### ISSUE TITLE
vagrant only: url redirect errors when accessing from a non-local machine

### ISSUE THREAD

**Issue by binary1230** [magfest/ubersystem#1377]
scenario: this is a weird interaction between using vagrant port forwarding + accessing the url on a different box + cherrypy's url redirect code.

repro:
spin up a vagrant instance on box A, let's say it's IP is 10.0.0.2

on box B (a physically separate box on the same network) go to https://10.0.0.2:4443/uber/

expected:  you get a redirect to `https://10.0.0.2:4443/uber`
actual: the web request will redirect you back to just `http://10.0.0.2/uber`

This is a bit of an edge case but does impact @earl7399's workflow, and may impact things when we start doing interesting stuff with port forwarding with the on-site ubersystem.

**Comment by binary1230** [comment]
I suspect this is nginx not rewriting the location correctly.

This line is supposed to do it in the nginx config, but I supect it's not handling all the cases.
```
proxy_redirect http://localhost/ $scheme://$host:$server_port/;
```

**Comment by binary1230** [comment]
to fix, we need the following in nginx:

```
    proxy_set_header      X-Real-IP $remote_addr;
    proxy_set_header      X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header      X-Forwarded-Host   $host:$server_port;
    proxy_set_header      X-Forwarded-Proto  $scheme;
```

That DEFINITELY fixes this IF all the port numbers are the same between the frontend and the backend.  Need to check the case where they aren't, like if we're forwarding traffic around.  Might need some slight adjustment.

With those enabled, we also need these lines in the ubersystem config:
```
tools.proxy.on = boolean(default=True)
tools.proxy.base = string(default="%(url_root)s")
```

**Comment by binary1230** [comment]
I am paying attention to this pre-event in case this year's config involves any kind of more magical proxying, and also because we should really just fix this anyway.

**Comment by binary1230** [comment]
OK! got it working even if we throw wacky port forwarding stuff at it.  For instance:

setup an SSH tunnel for a host arbitrarily named 'ubercool' on port 9999 and the forwarding endpoint is nginx listening on port 443.

when you get 301 redirects (i.e. the 'Location' HTTP header), nginx needs to return a url that starts with ```ubercool:9999``` without having that host:port explicitly named anywhere.

We can do this purely in nginx (which is mostly what we were doing before), or we can do it in cherrypy via the tools.proxy module.  Either way is OK, I just happened to get it working in cherrypy first.

TLDR for my future reference, if the nginx config is for 'labs.uber.magfest.org:4443' but the tunnel you access that through is 'ubercool:9999' then:

```$host:$server_port``` is 'labs.uber.magfest.org:4443' (pulled from nginx config)
```$host_http``` is the ```Host``` header passed in by the client which is 'ubercool:9999'.

---

The only thing that gets a little weird here is that the solution is perfectly valid, but now cherrypy knows it's running behind a proxy.  It can still function by being hit directly, but, normally the proxy (nginx) is the one that changes redirects like changing cherrypy responding with 'Location: http://localhost/uber/accounts/login' to make nginx say 'Location: http://labs.uber.magfest.org:443/uber/accounts/login'

Essentially our proxy_redirect directive in nginx is now unused because cherrypy is rewriting its own URLs based on info passed in.

If we ever decided we didn't want cherrypy to know it's behind a proxy, we'd simply not send the X-Forwarded-Host header to cherrypy from nginx.  And we'd probably have to modify nginx to say:

```
proxy_redirect        http://localhost/ $scheme://$http_host/;
```

then we would keep tools.proxy on in cherrypy (because totally unrelated to all this, we want to log the original real IP address of the user in cherrypy).  everything else should be good.

TLDR FINAL CONFIG:

INI file:
```
tools.proxy.on = True
```

```
    # proxy_redirect logic is currently being handled in cherrypy, but having this doesn't hurt.
    proxy_redirect        http://localhost/ $scheme://$http_host/;

    proxy_set_header      Host $host;
    proxy_set_header      X-Real-IP $remote_addr;
    proxy_set_header      X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header      X-Forwarded-Host   $http_host;   # 'tunnel_hostname:tunnel_port'
    proxy_set_header      X-Forwarded-Proto  $scheme;     # 'https', great so we don't redirect to http first
```

**Comment by binary1230** [comment]
also, this will very likely fix https://github.com/magfest/ubersystem/issues/1975 and https://github.com/RAMSProject/quickstart/issues/29

**Comment by binary1230** [comment]
fixed by https://github.com/magfest/ubersystem-puppet/pull/113
